src/group/mls_group/mod.rs

```diff
@@ -21,44 +21,39 @@ mod new_from_welcome;
 
 use crate::ciphersuite::*;
 use crate::codec::*;
-use crate::creds::*;
 use crate::framing::*;
 use crate::group::*;
 use crate::key_packages::*;
-use crate::messages::*;
+use crate::messages::{proposals::*, *};
 use crate::schedule::*;
-use crate::tree::astree::*;
-use crate::tree::*;
+use crate::tree::{astree::*, index::*, node::*, *};
 
 pub use api::*;
 use apply_commit::*;
 use create_commit::*;
 use new_from_welcome::*;
 
+use std::cell::{Ref, RefCell};
+
 pub struct MlsGroup {
     ciphersuite: Ciphersuite,
     group_context: GroupContext,
     generation: u32,
     epoch_secrets: EpochSecrets,
-    astree: ASTree,
-    tree: RatchetTree,
+    astree: RefCell<ASTree>,
```

Comment:
TODO: Have more than one ASTree and keep only the last n trees.

Reply:
I filed https://github.com/raphaelrobert/maelstrom/issues/16